### PR TITLE
Implement efficient_conv_bn_eval_decomp_graph_transform to handle conv and bn fusion after decomp

### PR DESCRIPTION
**Pull request by JackCaoG**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123805
* #123683
* #123673
* __->__ #123680

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by JackCaoG** [comment]
@ezyang I think this pr is ready for review. FIrst 3 pr is kind of no-op, the 4th pr will enable the inlining e2e. I will try to fix new test failures on 4th pr.

File: torch/_inductor/fx_passes/efficient_conv_bn_eval.py [hunk 1]
```diff
@@ -70,6 +75,164 @@ def efficient_conv_bn_eval(
...
+
+    # this node replaces the original conv + bn, and therefore
+    # should replace the uses of bn_node
+    bn_node.replace_all_uses_with(new_node)
+    # take care of the deletion order:
+    # delete bn_node first, and then conv_node
+    graph.erase_node(bn_node)
+    graph.erase_node(conv_node)
+
+    return
```
Comment (ezyang) [inline torch/_inductor/fx_passes/efficient_conv_bn_eval.py:233]: Needs a test, I guess?

**Comment by ezyang** [comment]
It's kind of hard to accept this PR without the code being used anywhere. Maybe we can delay this code until it actually gets used?

**Comment by JackCaoG** [comment]
yea, I think I separate this out mostly just for review purpose. I need to change this fn again after I enable the `CompositeImplicit decomps` since the pattern is different.

**Comment by JackCaoG** [comment]
Same for the first 3 commit in this stack, it doesn't make much sense to merge them separately. I separate them mostly to make review easier.

File: torch/_dynamo/config.py [hunk 2]
```diff
@@ -338,6 +338,9 @@ def _get_optimize_ddp_mode():
 # use numpy's PRNG if True, pytorch otherwise
 use_numpy_random_stream = False
 
+# (WIP)Generate a single graph for forward and backward.
+use_single_step_graph = False
```
Comment (ezyang) [inline torch/_dynamo/config.py:342]: this should be added by a later pr

Comment (JackCaoG) [reply torch/_dynamo/config.py:342]: sg, let me move.

**Comment by ezyang** [comment]
@youkaichao could you please help review this implementation?

**Comment by youkaichao** [comment]
Can I get some context on this change and the motivation?

**Comment by JackCaoG** [comment]
Hey @youkaichao I have a change int the stack https://github.com/pytorch/pytorch/pull/123683 which will decompose all nn modules with the config `use_signle_step_graph`. When that config is enabled, we won't see nn modules in the fx graph, but the ops like `torch.ops.aten.linear.default`. In this pr I replicate your fx pass to handle the fusion of the torch op versions of linear and conv.

**Comment by JackCaoG** [comment]
I didn't meant to delete your original stuff because my change is a bit too riskly. @ezyang and I think my change will likely be disabled in the head when it breaks other things, in that case I need both pass exists to handle the nn module version of the `linear` and op version of the `linear`.

In my replicated version of the pass
1. I used a `CallFunctionVarArgs` instead of `CallModuleVarArgs` since I need to match the `CallFunction` of the ops
2. `supported_convs` is updated to a list of ops instead of nn modules
3. `weight` and `bias`(as well as other paramteres) will not be the named parameter of the module anymore. They will be passed directly to the ops.

**Comment by JackCaoG** [comment]
I verified the correctness of my implementation by running `python test/inductor/test_efficient_conv_bn_eval.py EfficientConvBNEvalCpuTests.test_basic_cpu` after my third pr https://github.com/pytorch/pytorch/pull/123683 (which is where I find I break your test).

I thought about adding a test for this pass, however I also plan to enabled the compositeImplict decomp in dynamo, after that we won't see `linear` as torch ops, but we will see the decomposed version of linear(which I think is `matmul` + `transpose`). The test I added here won't be too useful..

**Comment by youkaichao** [comment]
Okay, so you didn't update my code, but duplicate my code and modify it. That's why I don't see the diff.

I think the general idea is clear, and you should test your implementation. It should be able to reuse tests in `test/inductor/test_efficient_conv_bn_eval.py`, by adding some flags.

**Comment by JackCaoG** [comment]
ok, let me see if I can manually decompose the modules before entering the dynamo to mimic what I plan to do.

File: torch/_inductor/fx_passes/efficient_conv_bn_eval.py [hunk 3]
```diff
@@ -70,6 +75,160 @@ def efficient_conv_bn_eval(
     return output
 
 
+def efficient_conv_bn_fn_eval(
```
Comment (youkaichao) [inline torch/_inductor/fx_passes/efficient_conv_bn_eval.py:78]: Quick question: does this shadow the original `efficient_conv_bn_fn_eval` function defined above?

Comment (JackCaoG) [reply torch/_inductor/fx_passes/efficient_conv_bn_eval.py:78]: It is copied and modified version of the `efficient_conv_bn_eval`, note that my copied version has a `fn` in the name so naming won't conflict.

**Comment by JackCaoG** [comment]
Test is updated, it turned out to be a lot easier than I thought :D.

File: torch/_inductor/fx_passes/efficient_conv_bn_eval.py (hunk 1 above)
Comment (JackCaoG) [reply torch/_inductor/fx_passes/efficient_conv_bn_eval.py:229]: added

File: torch/_inductor/fx_passes/efficient_conv_bn_eval.py (hunk 3 above)
Comment (youkaichao) [reply torch/_inductor/fx_passes/efficient_conv_bn_eval.py:78]: Got it.

Comment (youkaichao) [reply torch/_inductor/fx_passes/efficient_conv_bn_eval.py:78]: @JackCaoG I suggest suffixing `_fn` after the function name. Adding `_fn` in the function name is quite confusing.

Comment (JackCaoG) [reply torch/_inductor/fx_passes/efficient_conv_bn_eval.py:78]: let me renamed it to something like `efficient_conv_bn_eval_decomposed` to make it less confusing

**Comment by youkaichao** [comment]
@JackCaoG by the way, the paper title has changed to "Efficient ConvBN Blocks for Transfer Learning and Beyond" . Would you mind update it as well? Both in my original implementation and your decomposed implementation.

I didn't find time to update it. It would be great if you do in this PR.

**Comment by JackCaoG** [comment]
> @JackCaoG by the way, the paper title has changed to "Efficient ConvBN Blocks for Transfer Learning and Beyond" . Would you mind update it as well? Both in my original implementation and your decomposed implementation.
> 
> I didn't find time to update it. It would be great if you do in this PR.

sure

**Review by ezyang — APPROVED** [review]

**Review by youkaichao — APPROVED** [review]
Thanks, looks good to me! Is the failed test related to this PR? I cannot see the test failure log.

**Pull request closed without merging**